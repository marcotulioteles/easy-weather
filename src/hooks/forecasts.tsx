import { useContext, useEffect, createContext, ReactNode, useState, SetStateAction } from 'react';
import { oneCallWeatherDataAPI, directGeocodingAPI } from '../services/api';

type ForecastsProviderProps = {
  children: ReactNode;
}

type HourlyForecastData = {
  dt: number;
  temp: number;
  feels_like: number;
  humidity: number;
  clouds: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string;
    }
  ]
}

type DailyForecastData = {
  dt: number;
  temp: {
    day: number;
  };
  feels_like: {
    day: number;
  };
  humidity: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string
    }
  ];
  clouds: number;
}

type CurrentForecastData = {
  temp: number;
  feels_like: number;
  humidity: number;
  clouds: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string
    }
  ];
}

interface ForecastData {
  current: CurrentForecastData;
  hourly: HourlyForecastData[];
  daily: DailyForecastData[];
}

interface LocationDataResponse {
  name: string;
  country: string;
  state: string;
}

interface ForecastContextProps {
  locationInput: string;
  locationResponse: LocationDataResponse;
  setLocationInput: React.Dispatch<SetStateAction<string>>;
  error: boolean;
  forecast: ForecastData;
  loading: boolean;
}

export const ForecastsContext = createContext({} as ForecastContextProps);

function ForecastsProvider({ children }: ForecastsProviderProps) {
  const [locationInput, setLocationInput] = useState('');
  const [locationResponse, setLocationResponse] = useState<LocationDataResponse>({} as LocationDataResponse);
  const [forecast, setForecast] = useState<ForecastData>({} as ForecastData);
  const [error, setError] = useState(false);
  const [loading, setLoading] = useState(false);

  async function fetchForecast(locationInput: string) {
    try {
      setLoading(true);

      const response = await directGeocodingAPI
        .get(`direct?q=${locationInput}&appid=${process.env.NEXT_PUBLIC_APPID}`);
      const locationData = response.data;

      const locationFormatted = {
        name: locationData[0].name,
        country: locationData[0].country,
        state: locationData[0].state
      }

      setLocationResponse(locationFormatted);

      if (locationData) {
        const response = await oneCallWeatherDataAPI
          .get(`onecall?lat=${locationData[0].lat}&lon=${locationData[0].lon}&exclude=minutely,alerts&appid=${process.env.NEXT_PUBLIC_APPID}`)
        const forecastData = response.data;

        const dataFormatted = {
          current: {
            temp: forecastData.current.temp,
            feels_like: forecastData.current.feels_like,
            humidity: forecastData.current.humidity,
            clouds: forecastData.current.clouds,
            wind_speed: forecastData.current.wind_speed,
            weather: forecastData.current.weather,
          },
          hourly: forecastData.hourly.map((hour: HourlyForecastData) => {
            return {
              dt: hour.dt,
              temp: hour.temp,
              feels_like: hour.feels_like,
              humidity: hour.humidity,
              clouds: hour.clouds,
              wind_speed: hour.wind_speed,
              weather: hour.weather.map(weather => {
                return {
                  description: weather.description,
                  icon: weather.icon
                }
              })
            }
          }),
          daily: forecastData.daily.map((day: DailyForecastData) => {
            return {
              dt: day.dt,
              temp: day.temp,
              feels_like: day.feels_like,
              humidity: day.humidity,
              clouds: day.clouds,
              wind_speed: day.wind_speed,
              weather: day.weather.map(weather => {
                return {
                  description: weather.description,
                  icon: weather.icon
                }
              })
            }
          })
        }

        setForecast(dataFormatted);
        setLoading(false);
      }
    } catch (error) {
      setError(true);
    }
  }

  useEffect(() => {
    if (locationInput) fetchForecast(locationInput);
  }, [locationInput]);

  return (
    <ForecastsContext.Provider value={
      {
        locationInput,
        locationResponse,
        setLocationInput,
        error,
        forecast,
        loading
      }}>
      {children}
    </ForecastsContext.Provider>
  );
};

function useGetForecasts() {
  const context = useContext(ForecastsContext);

  return context;
}

export { ForecastsProvider, useGetForecasts };