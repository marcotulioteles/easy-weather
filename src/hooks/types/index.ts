export type HourlyForecastData = {
  dt: number;
  temp: number;
  feels_like: number;
  humidity: number;
  clouds: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string;
    }
  ]
}

export type DailyForecastData = {
  dt: number;
  temp: {
    min: number;
    max: number;
  };
  feels_like: {
    day: number;
  };
  humidity: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string
    }
  ];
  clouds: number;
}

export type CurrentForecastData = {
  temp: number;
  min_temp: number;
  max_temp: number;
  feels_like: number;
  humidity: number;
  clouds: number;
  wind_speed: number;
  weather: [
    {
      description: string;
      icon: string
    }
  ];
}

export interface ForecastData {
  current: CurrentForecastData;
  hourly: HourlyForecastData[];
  daily: DailyForecastData[];
}

export interface LocationData {
  name: string;
  country: string;
  state: string;
}