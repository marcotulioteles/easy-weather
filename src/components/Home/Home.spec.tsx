import { render } from '@testing-library/react';
import React from 'react';
import { HomeContent } from '.';
import { ForecastsContext } from '../../hooks/forecasts';
import { ForecastData, LocationData } from '../../hooks/types';
import { mockedForecastResponse, mockedLocationResponse } from '../../tests/__mocks__';

const useForecastValues = {
  locationInput: 'anyLocation',
  locationResponse: {} as LocationData,
  setLocationInput: jest.fn(),
  error: false,
  forecast: {} as ForecastData,
  loading: false,
  isEmpty: false
}

describe('<HomeContent />', () => {
  it('should render the component', () => {
    const component = render(
      <ForecastsContext.Provider value={useForecastValues}>
        <HomeContent />
      </ForecastsContext.Provider>
    );

    expect(component).toMatchSnapshot();
  });

  it('should render the empty state component if isEmpty', () => {
    render(
      <ForecastsContext.Provider value={
        { ...useForecastValues, isEmpty: true }
      }>
        <HomeContent />
      </ForecastsContext.Provider>
    );

    const emptyStateMessage = 'Your search did not find any data, please try again...';

    expect(emptyStateMessage).toBeDefined();
  });

  it('should render loading component', () => {
    const { getByAltText } = render(
      <ForecastsContext.Provider value={
        { ...useForecastValues, loading: true }
      }>
        <HomeContent />
      </ForecastsContext.Provider>
    );

    const loadingIconImage = getByAltText('weather-loading');

    expect(loadingIconImage).toBeInTheDocument();
  });

  it('should render forecast cards information', () => {
    const { getByText } = render(
      <ForecastsContext.Provider value={
        { ...useForecastValues, forecast: mockedForecastResponse, locationResponse: mockedLocationResponse }
      }>
        <HomeContent />
      </ForecastsContext.Provider>
    );

    const getTimeTitle = 'search time:';

    expect(getByText(getTimeTitle)).toBeInTheDocument();
  });
});