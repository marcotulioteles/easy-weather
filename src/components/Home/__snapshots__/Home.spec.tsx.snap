// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeContent /> should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="styles__Container-sc-lrxsm0-0 fotxIH"
      >
        <div
          class="styles__Container-sc-1ub8hmw-0 lhVrgj"
        >
          <div
            class="styles__LogoAndDate-sc-1ub8hmw-2 kzUXEJ"
          >
            <div
              class="styles__LogoContainer-sc-1ub8hmw-1 hBmjLk"
            >
              <span
                style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
              >
                <span
                  style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                >
                  <img
                    alt=""
                    aria-hidden="true"
                    src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2780%27%20height=%2780%27/%3e"
                    style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                  />
                </span>
                <img
                  alt="EasyWeatherLogo"
                  data-nimg="intrinsic"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
                <noscript />
              </span>
              <div
                class="styles__AppTitle-sc-1ub8hmw-3 iqlRYm"
              >
                <h1
                  class="styles__AppTitleFirstName-sc-1ub8hmw-4 krAcfM"
                >
                  Easy
                </h1>
                <p
                  class="styles__AppTitleLastName-sc-1ub8hmw-5 fsrlXR"
                >
                  Weather
                </p>
              </div>
            </div>
            <div
              class="styles__DateContainer-sc-1ub8hmw-6 ghaBpc"
            >
              <svg
                color="#2A9DC1"
                fill="none"
                height="28"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                style="color: rgb(42, 157, 193);"
                viewBox="0 0 24 24"
                width="28"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="4"
                />
                <line
                  x1="16"
                  x2="16"
                  y1="2"
                  y2="6"
                />
                <line
                  x1="8"
                  x2="8"
                  y1="2"
                  y2="6"
                />
                <line
                  x1="3"
                  x2="21"
                  y1="10"
                  y2="10"
                />
              </svg>
              <p
                class="styles__DateText-sc-1ub8hmw-7 laZUKz"
              >
                March 17, 2022 at 11:15 PM
              </p>
            </div>
          </div>
          <div
            class="styles__Location-sc-1ub8hmw-9 kstKDZ"
          />
          <form
            class="styles__ResearchContainer-sc-1ub8hmw-8 dcUwof"
          >
            <div
              class="styles__Container-sc-vzkx75-0 blYGbU"
            >
              <input
                class="styles__InputField-sc-vzkx75-1 bXUDHA"
                placeholder="Your location e.g.: (New York, US)"
                type="text"
                value=""
              />
              <svg
                color="#2A9DC1"
                fill="none"
                height="16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                style="color: rgb(42, 157, 193); position: absolute; left: 12px;"
                viewBox="0 0 24 24"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="11"
                  cy="11"
                  r="8"
                />
                <line
                  x1="21"
                  x2="16.65"
                  y1="21"
                  y2="16.65"
                />
              </svg>
            </div>
            <div
              class="styles__Container-sc-fbxr7z-0 McDWU"
            >
              <button
                class="styles__ButtonElement-sc-fbxr7z-1 iwRJDh"
                type="submit"
              >
                search
              </button>
            </div>
          </form>
        </div>
        <div
          class="styles__Container-sc-y41mfb-0 PPIfk"
        >
          <h1
            class="styles__Title-sc-y41mfb-1 gUPqFs"
          >
            FORECAST
          </h1>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="styles__Container-sc-lrxsm0-0 fotxIH"
    >
      <div
        class="styles__Container-sc-1ub8hmw-0 lhVrgj"
      >
        <div
          class="styles__LogoAndDate-sc-1ub8hmw-2 kzUXEJ"
        >
          <div
            class="styles__LogoContainer-sc-1ub8hmw-1 hBmjLk"
          >
            <span
              style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
            >
              <span
                style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2780%27%20height=%2780%27/%3e"
                  style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                />
              </span>
              <img
                alt="EasyWeatherLogo"
                data-nimg="intrinsic"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
            <div
              class="styles__AppTitle-sc-1ub8hmw-3 iqlRYm"
            >
              <h1
                class="styles__AppTitleFirstName-sc-1ub8hmw-4 krAcfM"
              >
                Easy
              </h1>
              <p
                class="styles__AppTitleLastName-sc-1ub8hmw-5 fsrlXR"
              >
                Weather
              </p>
            </div>
          </div>
          <div
            class="styles__DateContainer-sc-1ub8hmw-6 ghaBpc"
          >
            <svg
              color="#2A9DC1"
              fill="none"
              height="28"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="color: rgb(42, 157, 193);"
              viewBox="0 0 24 24"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                ry="2"
                width="18"
                x="3"
                y="4"
              />
              <line
                x1="16"
                x2="16"
                y1="2"
                y2="6"
              />
              <line
                x1="8"
                x2="8"
                y1="2"
                y2="6"
              />
              <line
                x1="3"
                x2="21"
                y1="10"
                y2="10"
              />
            </svg>
            <p
              class="styles__DateText-sc-1ub8hmw-7 laZUKz"
            >
              March 17, 2022 at 11:15 PM
            </p>
          </div>
        </div>
        <div
          class="styles__Location-sc-1ub8hmw-9 kstKDZ"
        />
        <form
          class="styles__ResearchContainer-sc-1ub8hmw-8 dcUwof"
        >
          <div
            class="styles__Container-sc-vzkx75-0 blYGbU"
          >
            <input
              class="styles__InputField-sc-vzkx75-1 bXUDHA"
              placeholder="Your location e.g.: (New York, US)"
              type="text"
              value=""
            />
            <svg
              color="#2A9DC1"
              fill="none"
              height="16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="color: rgb(42, 157, 193); position: absolute; left: 12px;"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="11"
                cy="11"
                r="8"
              />
              <line
                x1="21"
                x2="16.65"
                y1="21"
                y2="16.65"
              />
            </svg>
          </div>
          <div
            class="styles__Container-sc-fbxr7z-0 McDWU"
          >
            <button
              class="styles__ButtonElement-sc-fbxr7z-1 iwRJDh"
              type="submit"
            >
              search
            </button>
          </div>
        </form>
      </div>
      <div
        class="styles__Container-sc-y41mfb-0 PPIfk"
      >
        <h1
          class="styles__Title-sc-y41mfb-1 gUPqFs"
        >
          FORECAST
        </h1>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
